src/fsharp/infos.fs

```diff
@@ -161,18 +161,19 @@ type AllowMultiIntfInstantiations = Yes | No
 /// Visit base types and interfaces first.
 let private FoldHierarchyOfTypeAux followInterfaces allowMultiIntfInst skipUnref visitor g amap m typ acc = 
     let rec loop ndeep typ ((visitedTycon,visited:TyconRefMultiMap<_>,acc) as state) =
-
-        let seenThisTycon = isAppTy g typ && Set.contains (tcrefOfAppTy g typ).Stamp visitedTycon 
+        let isTypeApplication = isAppTy g typ
+        let tcrefOfTypeApplication = lazy (tcrefOfAppTy g typ)
```

Comment:
I can revert this part of the change if the lazy value here is costly, this function is recursive and is applied on type hierarchies.

Reply:
Definitely expensive